File: app/src/main/java/tw/firemaples/onscreenocr/ocr/OCRManager.java
package tw.firemaples.onscreenocr.ocr;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.os.AsyncTask;

import java.util.List;

public class OCRManager {
    private static OCRManager _instance;

    private Context context;

    private OnOCRStateChangedListener callback;
    private AsyncTask lastAsyncTask;

    private Bitmap currentScreenshot;
    private List<Rect> boxList;

    public static synchronized OCRManager getInstance(Context context) {
        if (_instance == null) {
            _instance = new OCRManager(context);
        }

        return _instance;
    }

    private OCRManager(Context context) {
        this.context = context;
    }

    public void setListener(OnOCRStateChangedListener callback) {
        this.callback = callback;
    }

    public void start(Bitmap screenshot, List<Rect> boxList) {
        this.currentScreenshot = screenshot;
        this.boxList = boxList;

        initOcrEngine();
    }

    public void cancelRunningTask() {
        if (lastAsyncTask != null) {
            lastAsyncTask.cancel(true);
        }
    }

    private void initOcrEngine() {
        if (callback != null) {
            callback.onInitializing();
        }

        lastAsyncTask = new OcrInitAsyncTask(context, onOcrInitAsyncTaskCallback).execute();
    }

    private OcrInitAsyncTask.OnOcrInitAsyncTaskCallback onOcrInitAsyncTaskCallback =
            new OcrInitAsyncTask.OnOcrInitAsyncTaskCallback() {
                @Override
                public void onOcrInitialized() {
                    if (callback != null) {
                        callback.onInitialized();
                    }
                    startTextRecognize();
                }

                @Override
                public void showMessage(String message) {
                }

                @Override
                public void hideMessage() {
                }
            };

    private void startTextRecognize() {
        if (callback != null) {
            callback.onRecognizing();
        }

        lastAsyncTask = new OcrRecognizeAsyncTask(context,
                currentScreenshot,
                boxList,
                onTextRecognizeAsyncTaskCallback).execute();
    }

    private OcrRecognizeAsyncTask.OnTextRecognizeAsyncTaskCallback onTextRecognizeAsyncTaskCallback =
            new OcrRecognizeAsyncTask.OnTextRecognizeAsyncTaskCallback() {
                @Override
                public void onTextRecognizeFinished(List<OcrResult> results) {
                    if (callback != null) {
                        callback.onRecognized(results);
                    }
                }

                @Override
                public void showMessage(String message) {
                }

                @Override
                public void hideMessage() {
                }
            };


    public interface OnOCRStateChangedListener {
        void onInitializing();

        void onInitialized();

        void onRecognizing();

        void onRecognized(List<OcrResult> results);
    }
}

File: app/src/main/java/tw/firemaples/onscreenocr/ocr/OCRManager.kt
package tw.firemaples.onscreenocr.ocr

import android.graphics.Bitmap
import android.graphics.Rect
import android.os.AsyncTask
import com.googlecode.tesseract.android.TessBaseAPI
import tw.firemaples.onscreenocr.utils.ImageFile
import tw.firemaples.onscreenocr.utils.Utils.Companion.context
import java.io.File

object OCRManager {
    val tessBaseAPI: TessBaseAPI by lazy { TessBaseAPI() }

    private var callback: OnOCRStateChangedListener? = null
    private var lastAsyncTask: AsyncTask<*, *, *>? = null

    private var currentScreenshot: ImageFile? = null
    private var boxList: List<Rect>? = null

    fun setListener(callback: OnOCRStateChangedListener) {
        this.callback = callback
    }

    fun start(screenshot: ImageFile, boxList: List<Rect>) {
        this.currentScreenshot = screenshot
        this.boxList = boxList

        initOcrEngine()
    }

    fun cancelRunningTask() {
        lastAsyncTask?.cancel(true)
    }

    private fun initOcrEngine() {
        callback?.onInitializing()

        lastAsyncTask = OcrInitAsyncTask(context, onOcrInitAsyncTaskCallback).execute()
    }

    private val onOcrInitAsyncTaskCallback = object :
            OcrInitAsyncTask.OnOcrInitAsyncTaskCallback {
        override fun onOcrInitialized() {
            callback?.onInitialized()
            startTextRecognize()
        }

        override fun showMessage(message: String) {}

        override fun hideMessage() {}
    }

    private fun startTextRecognize() {
        callback?.onRecognizing()

        lastAsyncTask = OcrRecognizeAsyncTask(context,
                currentScreenshot,
                boxList,
                onTextRecognizeAsyncTaskCallback).execute()
    }

    private val onTextRecognizeAsyncTaskCallback = object :
            OcrRecognizeAsyncTask.OnTextRecognizeAsyncTaskCallback {
        override fun onTextRecognizeFinished(results: List<OcrResult>) {
            callback?.onRecognized(results)
        }

        override fun showMessage(message: String) {}

        override fun hideMessage() {}
    }


    interface OnOCRStateChangedListener {
        fun onInitializing()

        fun onInitialized()

        fun onRecognizing()

        fun onRecognized(results: List<OcrResult>)
    }
}